### DISCUSSION TITLE
Current auto-instrumentation support for metrics

### DISCUSSION THREAD

**Discussion by SohamSankhe** [open-telemetry/opentelemetry-java-instrumentation#6419] [Q&A]
Hello,

I was exploring the use of Otel javaagent v1.16.0 for metrics and I could not find any documentation that lists all the available out of the box metrics. The document with 'list of supported third party libraries' seems to be mostly for tracing instrumentation.

Looking at the code, I see metrics related auto-instrumentation in below areas:
JVM runtime, system, DB connection pool, http client/server, rpc client/server, kafka clients.

Seeing that metrics was recently marked as Stable, I am assuming this is all the javaagent provides as of now.

Can you please confirm this?
And is there any documentation I can follow to track progress in this area for my future requirements.

**Comment by Roiocam** [comment]
i am not sure this answer is help... it is my own opinion

on document website [this page](https://opentelemetry.io/status/) mark metrics (api, sdk, protocol) is stable. that is not meaning there is many out of the box metrics for library.

In my understanding, this means that the APIs, SDKs for metrics collection are stable enough for Library developers or who want to impl observability to migrate the original solution to OpenTelemetry.

it is just like Docker Container Infrastructure is stable software, but you have to build your own image then you can run it.

**Comment by mateuszrzeszutek** [comment]
Hey @SohamSankhe ,

> I was exploring the use of Otel javaagent v1.16.0 for metrics and I could not find any documentation that lists all the available out of the box metrics. The document with 'list of supported third party libraries' seems to be mostly for tracing instrumentation.

The list of supported libraries contains all libs that we instrument in any way -- be it traces, metrics, or logs. Would it make sense for you if we added a column saying which telemetry signal an instrumentation produces?

**Reply by SohamSankhe** [reply]
Hi @mateuszrzeszutek,
I think that might help. Because my assumption is when metrics instrumentation would be added for, say tomcat, it would be added to /main/instrumentation/tomcat/ package itself where tracing instrumentation is in place. In that case, it would be helpful to know it produces metrics signals as well.

So, is there any documentation that lists currently supported auto-instrumented metrics?

**Reply by tuhao1020** [reply]
Are the metrics related to garbage collection turned on by default? I just find  some metrics like `process_runtime_jvm_xxxxx`, but I didn't find any related to garbage collection.

**Reply by mateuszrzeszutek** [reply]
> Are the metrics related to garbage collection turned on by default?

Yes -- these are currently not specified in the [OTel semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/runtime-environment-metrics.md#jvm-metrics), and this cannot be turned on by default. You can enable them by setting `otel.instrumentation.runtime-metrics.experimental-metrics.enabled` to true.

**Reply by tuhao1020** [reply]
I have set `-Dotel.instrumentation.runtime-metrics.experimental-metrics.enabled=true` but still no gc metrics. `process.runtime.jvm.buffer.*` also missing.

```
# TYPE process_runtime_jvm_memory_committed gauge
# HELP process_runtime_jvm_memory_committed Measure of memory committed
process_runtime_jvm_memory_committed{pool="PS Old Gen",type="heap"} 5.0855936E7 1661961126455
process_runtime_jvm_memory_committed{pool="PS Survivor Space",type="heap"} 9961472.0 1661961126455
process_runtime_jvm_memory_committed{pool="Metaspace",type="non_heap"} 5.3346304E7 1661961126455
process_runtime_jvm_memory_committed{pool="PS Eden Space",type="heap"} 2.91504128E8 1661961126455
process_runtime_jvm_memory_committed{pool="Code Cache",type="non_heap"} 1.3959168E7 1661961126455
process_runtime_jvm_memory_committed{pool="Compressed Class Space",type="non_heap"} 7208960.0 1661961126455
# TYPE process_runtime_jvm_classes_unloaded_total counter
# HELP process_runtime_jvm_classes_unloaded_total Number of classes unloaded since JVM start
process_runtime_jvm_classes_unloaded_total 0.0 1661961126455
# TYPE process_runtime_jvm_classes_loaded_total counter
# HELP process_runtime_jvm_classes_loaded_total Number of classes loaded since JVM start
process_runtime_jvm_classes_loaded_total 9621.0 1661961126455
# TYPE process_runtime_jvm_classes_current_loaded gauge
# HELP process_runtime_jvm_classes_current_loaded Number of classes currently loaded
process_runtime_jvm_classes_current_loaded 9621.0 1661961126455
# TYPE process_runtime_jvm_system_cpu_load_1m gauge
# HELP process_runtime_jvm_system_cpu_load_1m Average CPU load of the whole system for the last minute
process_runtime_jvm_system_cpu_load_1m 0.8681640625 1661961126455
# TYPE process_runtime_jvm_threads_count gauge
# HELP process_runtime_jvm_threads_count Number of executing threads
process_runtime_jvm_threads_count 28.0 1661961126455
# TYPE process_runtime_jvm_cpu_utilization gauge
# HELP process_runtime_jvm_cpu_utilization Recent cpu utilization for the process
process_runtime_jvm_cpu_utilization 0.0 1661961126455
# TYPE process_runtime_jvm_memory_limit gauge
# HELP process_runtime_jvm_memory_limit Measure of max obtainable memory
process_runtime_jvm_memory_limit{pool="PS Old Gen",type="heap"} 6.87865856E8 1661961126455
process_runtime_jvm_memory_limit{pool="PS Survivor Space",type="heap"} 9961472.0 1661961126455
process_runtime_jvm_memory_limit{pool="PS Eden Space",type="heap"} 3.21912832E8 1661961126455
process_runtime_jvm_memory_limit{pool="Code Cache",type="non_heap"} 2.5165824E8 1661961126455
process_runtime_jvm_memory_limit{pool="Compressed Class Space",type="non_heap"} 1.073741824E9 1661961126455
# TYPE process_runtime_jvm_memory_init gauge
# HELP process_runtime_jvm_memory_init Measure of initial memory requested
process_runtime_jvm_memory_init{pool="PS Old Gen",type="heap"} 4.3515904E7 1661961126455
process_runtime_jvm_memory_init{pool="PS Survivor Space",type="heap"} 2621440.0 1661961126455
process_runtime_jvm_memory_init{pool="Metaspace",type="non_heap"} 0.0 1661961126455
process_runtime_jvm_memory_init{pool="PS Eden Space",type="heap"} 1.6252928E7 1661961126455
process_runtime_jvm_memory_init{pool="Code Cache",type="non_heap"} 2555904.0 1661961126455
process_runtime_jvm_memory_init{pool="Compressed Class Space",type="non_heap"} 0.0 1661961126455
# TYPE process_runtime_jvm_memory_usage gauge
# HELP process_runtime_jvm_memory_usage Measure of memory used
process_runtime_jvm_memory_usage{pool="PS Old Gen",type="heap"} 1.5731704E7 1661961126455
process_runtime_jvm_memory_usage{pool="PS Survivor Space",type="heap"} 9949248.0 1661961126455
process_runtime_jvm_memory_usage{pool="Metaspace",type="non_heap"} 5.0121E7 1661961126455
process_runtime_jvm_memory_usage{pool="PS Eden Space",type="heap"} 4.9682264E7 1661961126455
process_runtime_jvm_memory_usage{pool="Code Cache",type="non_heap"} 1.3379136E7 1661961126455
process_runtime_jvm_memory_usage{pool="Compressed Class Space",type="non_heap"} 6616072.0 1661961126455
# TYPE process_runtime_jvm_system_cpu_utilization gauge
# HELP process_runtime_jvm_system_cpu_utilization Recent cpu utilization for the whole system
process_runtime_jvm_system_cpu_utilization 0.0196078431372549 1661961126455
# TYPE queueSize gauge
# HELP queueSize The number of spans queued
queueSize{spanProcessorType="BatchSpanProcessor"} 0.0 1661961126455
```

**Reply by mateuszrzeszutek** [reply]
What javaaagent version are you using?

**Reply by tuhao1020** [reply]
latest 1.17.0

**Reply by mateuszrzeszutek** [reply]
Hmm, that's weird, you should be seeing these. Can you open a bug for this? Please post your full javaagent config, the JVM used, and the full output of the prometheus exporter; also, it'd be great if you could run the javaagent in the debug mode (`-Dotel.javaagent.debug=true`) and attach the full logs as well.

**Reply by tuhao1020** [reply]
@mateuszrzeszutek #6532 
seems no metrics for daemon threads too

**Reply by gabrielgiussi** [reply]
@tuhao1020 I saw you did a mistake and gc was working fine for you after fixing that, what was it?

**Reply by tuhao1020** [reply]
@gabrielgiussi it's working now

**Reply by gabrielgiussi** [reply]
But did you change something like bumping the version or something like that?

**Reply by tuhao1020** [reply]
I do nothing and it was back to normal 😄

**Comment by gabrielgiussi** [comment]
@mateuszrzeszutek 

> Would it make sense for you if we added a column saying which telemetry signal an instrumentation produces?

It think it would be nice, I'm not seeing any kafka/jetty metrics after adding the agent but I don't know if they are not provided or it is something else in my configuration (because I'm also not seeing gc metrics and I understand those are exported).
I do see some metrics collected by the agent, e.g.
```
process_runtime_jvm_cpu_utilization
process_runtime_jvm_memory_committed
process_runtime_jvm_threads_count
http_client_response_size
http_client_duration
```
